### ISSUE TITLE
Migrate to Debian and Restore Failed - Help Please!

### ISSUE THREAD

**Issue by mr-handerson** [TechPerplexed/Gooby#170]
Going to need some assistance please... sigh. 

I should have just left everything as it was. Everything has been working wonderfully for months and months. Except I was running on Ubuntu 18.04, I finally had a little downtime, so I thought I would make the jump over to Debian. I've now been tearing my hair out for about 6 hours and getting nowhere fast.

Have a Full Backup to work with from yesterday. I thought that would be all I needed.

For each attempt I'm doing the following:
Server wipe and Debian 10 (Buster) install. Connect via SSH. Change to root. Install Gooby. Install rclone. 
This much seems successful.

Attempt 1:
Full restore. Much sadness. It didn't work at all and there were so many errors I didn't even know where to start.

Attempt 2:
Try and restore a single app (without installing via Gooby beforehand). Errors galore, lots of permission denied.

Attempt 3:
Confirm that the Gooby created user is a member of sudo before installing rclone (it is). Install rclone.
Set the timezone
Do a System Cleanup - this is successful. Gooby now has a version 2.2.2 listed.
Install a single app via Gooby first
Then try and restore the single app. Errors, see below.

```
Copying Ombi backup from GDrive...

3.120M / 3.120 MBytes, 100%, 173.373 kBytes/s, ETA 0s
rm: cannot remove '/var/local/Gooby/Docker/.config/snapshots/Ombi.snar': No such file or directory

 Ombi backup downloaded, proceeding...


 Taking containers down...
/bin/sh: 1: cannot create /dev/null: Permission denied
Stopping letsencrypt ... error
Stopping ombi        ... error
Stopping autoheal    ... error
Stopping watchtower  ... error
Stopping nginx-proxy ... error

ERROR: for autoheal  cannot stop container: 482388afad45fb212a357e60bdf8d2381c868f5e2ad70e19e18e62a9d3e63cb7: Cannot kill container 482388afad45fb212a357e60bdf8d2381c868f5e2ad70e19e18e62a9d3e63cb7: unknown error after kill: open /dev/null: no such file or directory: <nil>: unknown

ERROR: for letsencrypt  cannot stop container: 36171ad4d404fc9b62124f3995fcfecb3cff38077da7c97871cdbf596d9511fa: Cannot kill container 36171ad4d404fc9b62124f3995fcfecb3cff38077da7c97871cdbf596d9511fa: unknown error after kill: open /dev/null: no such file or directory: <nil>: unknown

ERROR: for ombi  cannot stop container: 99613ecd1b3397bc1a03527ff6db954bd0294ad17be8a2f024157310640737ed: Cannot kill container 99613ecd1b3397bc1a03527ff6db954bd0294ad17be8a2f024157310640737ed: unknown error after kill: open /dev/null: no such file or directory: <nil>: unknown

ERROR: for watchtower  cannot stop container: 7603d2d63185ff944ad2b84b0b0fdcf0cb0cb8d03f5ce6bade71723551c04448: Cannot kill container 7603d2d63185ff944ad2b84b0b0fdcf0cb0cb8d03f5ce6bade71723551c04448: unknown error after kill: open /dev/null: no such file or directory: <nil>: unknown
Removing network docker_default
ERROR: error while removing network: network docker_default id bb98d698f699634aeadc88b261cca688412e83e80a6548f4290b8808dee17363 has active endpoints

 Restoring files...

 Extracting full archive... /tmp/goobyrestore/Ombi-full.tar.gz

<SNIP>

 Extracting differential archive... /tmp/goobyrestore/*-diff.tar.gz


rm: cannot remove '/var/local/Gooby/Docker/.config/snapshots/*': No such file or directory

Incremental backups reset to full backup

/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
/opt/Gooby/install/misc/environment-build.sh: line 16: /dev/null: Permission denied
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    12  100    12    0     0    342      0 --:--:-- --:--:-- --:--:--   342
done
/bin/sh: 1: cannot create /dev/null: Permission denied
Pulling letsencrypt-nginx-proxy-companion (jrcs/letsencrypt-nginx-proxy-companion:)...
latest: Pulling from jrcs/letsencrypt-nginx-proxy-companion
540db60ca938: Already exists
0db47e060084: Already exists
e14b93b82599: Already exists
24999acb704a: Already exists
fd88cd68b2e4: Already exists
dc132704d121: Already exists
Digest: sha256:4eed9ea30bbecd4980e0c649e1f3320439de98ab8a851fae0ec97979c027bd3b
Status: Downloaded newer image for jrcs/letsencrypt-nginx-proxy-companion:latest
autoheal is up-to-date
watchtower is up-to-date
nginx-proxy is up-to-date
ombi is up-to-date
letsencrypt is up-to-date
 Restoring permissions, please wait...


 Finished restoring Ombi


 Make sure you check if your services are
 running properly before you remove the old installation!

 Remove old installation files (Y/n)? y

--------------------------------------------------
 restore backup completed
--------------------------------------------------

/opt/Gooby/install/misc/backup-restore.sh: line 165: /dev/null: Permission denied

--------------------------------------------------
 All done! Press Enter to return to the menu
--------------------------------------------------

```
Attempt 4: Starting a wipe and reinstall and I don't know what to try next
...
So I ended up trying putting 18.04 back on. Full Restore seemed to complete OK. Assuming I hit rate limit on letsencrypt as all my certificates were missing. However, once I tried to do an rclean and a reboot everything failed. Didn't recognize that any apps were installed. Could not manually install apps. Tried an uninstall and reinstall of rclone, which worked but didn't fix anything. Further attempts at restoring failed (forgot to get the error messages).

So I'm wiping again but will have to call it a night soon.

**Comment by TechPerplexed** [comment]
Ugh, sounds like you had a hard time of it. I found that Debian can display a lot of errors at times, but... after a reboot everything seems okay anyway. I know it's a bit late to suggest that now ^_^

**Comment by mr-handerson** [comment]
It's been very frustrating to say the least hah. Thanks for checking in.

Unfortunately my downtime disappeared today, so I'm only just able to get back to working on this now but I've already run into issues.

So I did a fresh Debian 10 install. 
Installed Gooby as root. 
Confirmed the created user is a member of sudo. 
Set the timezone. 
Installed rclone including using my client id and secret. 
Confirmed that autoheal, watchtower, nginx and letsencrypt containers were running and healthy. 
Reboot server.
No apps listed as installed or running now when I check
/mnt only has local and uploads folders
Removed rclone and got the following 

```
/bin/fusermount: entry for /mnt/rclone not found in /etc/mtab
rm: cannot remove '/usr/local/share/man/man1/rclone.1': No such file or directory
Removed /etc/systemd/system/default.target.wants/mergerfs.service.
Removed /etc/systemd/system/default.target.wants/rclonefs.service.
Removed /etc/systemd/system/mergerfs.service.wants/rclonefs.service.
Removed /etc/systemd/system/unionfs.service.wants/rclonefs.service.
```

So something seems amiss right from the outset, but unfortunately I don't have sufficient knowledge to properly troubleshoot.

Any ideas?

In the meantime I will play around with trying to get rclone installed and happy since that is a must-have.

**Comment by mr-handerson** [comment]
So after a reboot there are still no apps showing, and when I go to try and install rclone again I get this

```
--------------------------------------------------
 Rclone appears to be installed
--------------------------------------------------

Here is a list of the root folders in your mount:

ls: cannot access '/mnt/google': No such file or directory

If you're getting an error here, please
check your Rclone config settings.
You may need to uninstall and try again!
```
Trying to remove rclone again results in this
```
--------------------------------------------------
 Starting remove Rclone
--------------------------------------------------

/bin/fusermount: entry for /mnt/rclone not found in /etc/mtab
rm: cannot remove '/usr/local/share/man/man1/rclone.1': No such file or directory

--------------------------------------------------
 remove Rclone completed
--------------------------------------------------
```

So I'm stuck. I don't know what I'm doing wrong or how to proceed.

**Comment by mr-handerson** [comment]
So I decided just to jump right in, and I tried a full restore again.
It seemed to complete without errors!
Everything seems to be back... to a degree

It looks like I now have some out-of-date yamls. Do I have to manually edit each one to the most recent version, or is there some way to force a pull from git?

Also, portainer is showing as #64 for me, but is now listed as #14 in git. Can I just edit the number? Or will that break things? If I do manage to get things back up and running (I really hope so!) I'll be kind of hesitant to mess with things further. But at the same time I'd like to have everything current and correct so that going forwards it should continue to work with any subsequent updates.

**Comment by TechPerplexed** [comment]
To address your first questions: yes a Gooby restore really restores the entire environment, so if there are updated yaml files, you would have to manually change them. You can safely update them in the components folder (rename, edit content, etc) and then run `rclean` - you shouldn't have any issues that way.

As for your Rclone issue, that sounds a little more complicated... the simple answer is that if Rclone doesn't come up, none of your containers will be created either as a failsafe (especially Plex tends to delete all content if it can't find a mount). That is the easy part 😄 But now the question why Rclone doesn't play nice and I suggest we troubleshoot that before anything else.

If you follow these steps:

sudo systemctl stop mergerfs
sudo systemctl stop rclonefs

/bin/fusermount -uz /mnt/rclone
/bin/fusermount -uz /mnt/google

(at this point check if /mnt/rclone and /mnt google are gone from your server)

sudo systemctl daemon-reload
sudo systemctl start rclonefs
sudo systemctl start mergerfs

Do you get any errors along the way?

**Comment by mr-handerson** [comment]
Firstly, thanks again for the ongoing assistance! Very much appreciated.
> 
> 
> To address your first questions: yes a Gooby restore really restores the entire environment, so if there are updated yaml files, you would have to manually change them. You can safely update them in the components folder (rename, edit content, etc) and then run `rclean` - you shouldn't have any issues that way.
> 
OK, that makes sense. I have updated them as appropriate and will rclean once rclone is sorted out and the containers are all back.

As a follow-up question: If a yaml is updated (or any file for that matter I guess) in the master branch, how is that change pulled to our deployments? Is that what the Gooby update function is for? 

>
> As for your Rclone issue, that sounds a little more complicated... the simple answer is that if Rclone doesn't come up, none of your containers will be created either as a failsafe (especially Plex tends to delete all content if it can't find a mount). That is the easy part 😄 But now the question why Rclone doesn't play nice and I suggest we troubleshoot that before anything else.
> 
> If you follow these steps:
> 
> sudo systemctl stop mergerfs
> sudo systemctl stop rclonefs
> 
> /bin/fusermount -uz /mnt/rclone

/bin/fusermount: entry for /mnt/rclone not found in /etc/mtab
> /bin/fusermount -uz /mnt/google

/bin/fusermount: entry for /mnt/google not found in /etc/mtab

> 
> (at this point check if /mnt/rclone and /mnt google are gone from your server)

They are gone
> 
> sudo systemctl daemon-reload
> sudo systemctl start rclonefs
> sudo systemctl start mergerfs
> 
> Do you get any errors along the way?

Only errors were the 2 listed above. After completing, /mnt/rclone and /mnt/google reappeared

**Comment by TechPerplexed** [comment]
> As a follow-up question: If a yaml is updated (or any file for that matter I guess) in the master branch, how is that change pulled to our deployments? Is that what the Gooby update function is for?

The short answer is: it doesn't :) The reason why we don't update yaml files is not to break custom edits by users, so we leave it well alone. We don't actually ever replace any of your files in /var/local/Gooby, except in very, very rare cases and even then it would be in an official release. What you are referring to is pulling Gooby to /opt/Gooby, which is where most of the cron scripts dwell. Lately we haven't had to make many tweaks, so pulling it each time you run rclean isn't quite as necessary any longer, but it doesn't hurt anything either so it's still there.

> Only errors were the 2 listed above. After completing, /mnt/rclone and /mnt/google reappeared

That's how it is supposed to be! So let's leave out the forced unmount:

sudo systemctl stop mergerfs
sudo systemctl stop rclonefs

(check if /mnt/rclone and /mnt/google are there)

sudo systemctl daemon-reload
sudo systemctl start rclonefs
sudo systemctl start mergerfs

(check if /mnt/rclone and /mnt/google are back in place)

Still no errors?

**Comment by mr-handerson** [comment]
> 
> 
> > As a follow-up question: If a yaml is updated (or any file for that matter I guess) in the master branch, how is that change pulled to our deployments? Is that what the Gooby update function is for?
> 
> The short answer is: it doesn't :) The reason why we don't update yaml files is not to break custom edits by users, so we leave it well alone. We don't actually ever replace any of your files in /var/local/Gooby, except in very, very rare cases and even then it would be in an official release. What you are referring to is pulling Gooby to /opt/Gooby, which is where most of the cron scripts dwell. Lately we haven't had to make many tweaks, so pulling it each time you run rclean isn't quite as necessary any longer, but it doesn't hurt anything either so it's still there.
> 

OK. That also makes sense not wanting to overwrite custom edits.

But in that case we just need to keep an eye on changes that may actually need to be implemented, and manually make them? Eg. changes to which Docker image is used, or in the case of the proxy yaml, the addition of the acme.sh line

And that would only impact yaml's? Any updates to  cron scripts or elsewhere would get pulled during an update?

> > Only errors were the 2 listed above. After completing, /mnt/rclone and /mnt/google reappeared
> 
> That's how it is supposed to be! So let's leave out the forced unmount:
> 
> sudo systemctl stop mergerfs
> sudo systemctl stop rclonefs
> 
> (check if /mnt/rclone and /mnt/google are gone)

They are gone.
> 
> sudo systemctl daemon-reload
> sudo systemctl start rclonefs
> sudo systemctl start mergerfs
> 
> (check if /mnt/rclone and /mnt/google are back in place)

They are back
> 
> Still no errors?

No errors this time

**Comment by TechPerplexed** [comment]
That is promising :) At least we know it isn't anything to do with Rclone then. What we did was just force it to shut down properly and I had a feeling it might be back to normal after that!

As for your new error and your rate limit issue, I think it might be related to #144 so Please you might want to make sure that your /var/local/Gooby/Docker/docker-compose.yaml file  on line 41 includes the line
- ${CONFIGS}/Docker/acme.sh:/etc/acme.sh.

That should solve your rate problem... at least within the next few days!

**Comment by mr-handerson** [comment]
> 
> 
> That is promising :) At least we know it isn't anything to do with Rclone then. What we did was just force it to shut down properly and I had a feeling it might be back to normal after that!
> 
Indeed. Albeit odd too. After initially rushing in and failing spectacularly I was more methodical in my subsequent attempts but had been unable to even get rclone stable. Those few commands seemed to do the trick though.

Whatever the case, I am extremely grateful for the help! It was a very EEK moment when everything went pear shaped, and I am rather out of my depth in the *nix world.

>
> As for your new error and your rate limit issue, I think it might be related to #144 so you might want to make sure that your 03-proxy.yaml file includes the line
> 
> ` - ${CONFIGS}/Docker/acme.sh:/etc/acme.sh`
> 
> That should solve your rate problem... at least within the next few days!

I copied in clean copies of all the yaml's. I just double checked and both proxy and docker_compose contain the line referenced above. So if I understand correctly from #144 now I just need to wait it out until the limit resets? ..........

In checking everything over today it seems like things are otherwise back to normal. Things are automatically being downloaded and synced across. Everything seems to be updating. HUZZAH!!

One small item to note: Because I was using the default path, it seems my NZBGET scripts were not being backed up. If anyone else uses them, it might be worthwhile to change the path to a location that is included in the backups (eg. /config perhaps). Or make sure to manually make a copy.

**Comment by TechPerplexed** [comment]
Yay 😄 Keep us posted!!

**Comment by mr-handerson** [comment]
So one thing I hadn't tested yet was actually watching something.
It does not work. I initially though it might be because of the certs issue? And I just have to wait it out?
But I realised that I also don't have hardware transcoding setup properly again, and I'm struggling to figure out how in Debian. A lot of what I've found so far has been for Ubuntu and it doesn't seem to work for Debian, and I'm afraid of making a mess of things.
For starters I don't seem to currently have a /dev/dri directory
I should note that hardware transcoding did work previously in my Ubuntu installation

I'll keep researching and reading up but again would appreciate some guidance from someone who actually knows what they're doing.

**Comment by mr-handerson** [comment]
Hey no problem. I wasn't aware either, so we're all learning as we go. I'm stubborn so I'll keep at it for a while longer yet and will post if I'm successful or if I throw in the towel and go back to Ubuntu.

I did manage to find an old reddit thread that seems to be the exact issue I'm having, but unfortunately there was no final update as to whether they were ultimately successful in their attempts. I might try reaching out to see if the user recalls what happened in the end.

And yes, I will also try hitting some other forums to see if anyone can assist.

Thanks as always for your assistance! Other than still waiting on certs it seems I am back up and running. I can watch, I'd just prefer to be able to use hw transcoding since I definitely notice the difference in utilization between the two.

Once my certs have all updated I'll close this issue out.

**Comment by mr-handerson** [comment]
I've identified a folder permissions issue that I'm trying to figure out, but maybe someone can chime in and save me some time.

When my server install was completed, a user 'debian' was automatically created. I used it to login but then created a different user when setting up Gooby.
When the server is restarted the /tmp/Transcode directory used by Plex is removed, and then recreated when Plex starts.
The owner is the 'debian' account and it's created with 755 permissions. As such, I get access denied errors and can't watch anything.
If I manually chown or chmod the issue is resolved, until next reboot.

How do I best fix it so that the permissions on /tmp/Transcode are set properly upon creation?
